File: privacy_local_agent/routers/qol.py
"""查询混淆路由（qol/obfuscate|batch）。"""

from fastapi import APIRouter

from ..deps import SECURITY_DEPS, handle_request_exception, service
from ..schemas import QolBatchRequest, QolRequest
from ..security.auth import require_permission

router = APIRouter()


@router.post("/v1/privacy/qol/obfuscate", dependencies=[*SECURITY_DEPS, require_permission("privacy:qol")])
def qol_obfuscate(req: QolRequest):
    """查询混淆接口。

    Args:
        req: QolRequest 请求体。

    Returns:
        {"result": <混淆后的查询列表>}
    """
    return {"result": service.obfuscate_query(
        req.query,
        req.num_dummies,
        req.domain,
        medical_pool=req.medical_pool,
        generic_pool=req.generic_pool,
        seed=req.seed,
    )}


@router.post("/v1/privacy/qol/obfuscate/batch", dependencies=[*SECURITY_DEPS, require_permission("privacy:qol")])
def qol_obfuscate_batch(req: QolBatchRequest):
    """批量查询混淆接口。"""
    try:
        return {"results": service.obfuscate_query_batch(
            req.queries,
            req.num_dummies,
            req.domain,
            medical_pool=req.medical_pool,
            generic_pool=req.generic_pool,
            seed=req.seed,
        )}
    except Exception as e:
        handle_request_exception(e)

File: console/backend-go/internal/mapper/qol.go
package mapper

import (
	"context"
	"encoding/json"

	pb "github.com/fengzhizi319/privacy-local-agent/console/backend-go/proto"
)

// ---------------------------------------------------------------------------
// Query Obfuscation handlers —— 查询混淆
//
// 向真实查询中注入虚假查询（dummy queries），
// 使攻击者（如数据库管理员、网络监控者）无法区分哪些是真实查询。
// ---------------------------------------------------------------------------

// handleObfuscateQuery 处理 /v1/privacy/qol/obfuscate 路径，单条查询混淆。
//
// 在真实查询旁注入 num_dummies 个虚假查询，返回混合后的查询列表。
func (m *Mapper) handleObfuscateQuery(ctx context.Context, client pb.PrivacyServiceClient, body json.RawMessage) (any, error) {
	v, err := decode(body)
	if err != nil {
		return nil, err
	}
	// 构造 ObfuscateQueryRequest：query 为真实查询，num_dummies 为虚假查询数量
	resp, err := client.ObfuscateQuery(ctx, &pb.ObfuscateQueryRequest{
		Query:        getString(v, "query", ""),              // 真实查询
		NumDummies:   getInt32(v, "num_dummies", 3),          // 虚假查询数量（默认 3）
		Domain:       getString(v, "domain", "medical"),      // 查询领域（"medical"/"generic"）
		MedicalPool:  getStrings(v, "medical_pool"),          // 自定义医疗领域虚假查询池
		GenericPool:  getStrings(v, "generic_pool"),          // 自定义通用领域虚假查询池
	})
	if err != nil {
		return nil, err
	}
	// 返回混合后的查询列表（真实 + 虚假，顺序已打乱）
	return map[string][]string{"result": resp.Result}, nil
}

// handleObfuscateQueryBatch 处理 /v1/privacy/qol/obfuscate/batch 路径，批量查询混淆。
//
// 对多条查询分别注入虚假查询，返回每条查询的混淆结果。
func (m *Mapper) handleObfuscateQueryBatch(ctx context.Context, client pb.PrivacyServiceClient, body json.RawMessage) (any, error) {
	v, err := decode(body)
	if err != nil {
		return nil, err
	}
	// 构造 ObfuscateQueryBatchRequest：queries 为多条真实查询
	resp, err := client.ObfuscateQueryBatch(ctx, &pb.ObfuscateQueryBatchRequest{
		Queries:      getStrings(v, "queries"),             // 多条真实查询
		NumDummies:   getInt32(v, "num_dummies", 3),        // 每条查询的虚假查询数量
		Domain:       getString(v, "domain", "medical"),    // 查询领域
		MedicalPool:  getStrings(v, "medical_pool"),        // 自定义医疗虚假查询池
		GenericPool:  getStrings(v, "generic_pool"),        // 自定义通用虚假查询池
	})
	if err != nil {
		return nil, err
	}
	// 将 protobuf 重复消息列表展平为 [][]string
	results := make([][]string, 0, len(resp.Results))
	for _, r := range resp.Results {
		// 每个 r.Result 为一条查询的混淆结果（真实 + 虚假）
		results = append(results, r.Result)
	}
	return map[string][][]string{"results": results}, nil
}
